minorminer/utils/feasibility.py

```diff
@@ -0,0 +1,222 @@
+    by exploiting additional graph structure.
+
+    Args:
+        S: The source graph to embed.
+        T: The target graph in which to embed.
+        one_to_one: If True, only 1-to-1 (subgraph) embeddings are allowed.
+            Defaults to False, permitting minor embeddings.
+
+    Returns:
+        bool: `False` if subgraph embedding is definitely infeasible, `True`
+            otherwise.
+    """
+    if T == S:
+        return True
+    if (
+        T.number_of_nodes() < S.number_of_nodes()
+        or T.number_of_edges() < S.number_of_edges()
+    ):
+        return False
+    else:
+        S_degree = np.sort([S.degree[n] for n in S.nodes()])
+        T_degree = np.sort([T.degree[n] for n in T.nodes()])
+
+        if np.any(T_degree[-len(S_degree) :] < S_degree):
+            if one_to_one or T_degree[-1] <= 2:
+                # Too many high degree nodes in S
+                return False
+            else:  # Attempt minor embed (enhance T degrees)
+                # Minor embedding feasibility reduces to bin packing when
+                # considering a best target case knowing only the degree
+                # distribution. In general feasibility is NP-complete, a cheap
+                # marginal degree distribution filter is used.
+
+                # We can eliminate nodes of equal degree assuming best case:
+                ResidualCounts = Counter(T_degree)
+                ResidualCounts.subtract(Counter(S_degree))
+
+                # Target nodes of degree x <=2 are only of use for minor
+                # embedding source nodes of degree <=x:
+                for kS in range(3):
+                    if ResidualCounts[kS] < 0:
+                        ResidualCounts[kS + 1] += ResidualCounts[kS]
+                    ResidualCounts[kS] = 0
+
+                if all(v > 0 for v in ResidualCounts.values()):
+                    return True
+                nT_auxiliary = sum(ResidualCounts.values())
+                if nT_auxiliary < 0:  # extra available to form chains
+                    return False
+
+                # In best case all target nodes have degree kTmax, and chains
+                # are trees. To cover degree k in S requires n auxiliary target
+                # nodes such that kTmax + n(kTmax-2) >= k
+                kTmax = np.max([k for k, v in ResidualCounts.items() if v > 0])
+                min_auxiliary_necessary = sum(
+                    [
+                        -v * np.ceil((k - kTmax) / (kTmax - 2))
+                        for k, v in ResidualCounts.items()
+                        if v < 0
+                    ]
+                )
+                return min_auxiliary_necessary <= nT_auxiliary
+        else:
+            return True
+
+
+def lattice_size_upper_bound(T: nx.Graph = None) -> int:
+    """Determines an upper bound on the lattice size.
+
+    The lattice size is the parameter ``m`` of a dwave_networkx graph, also
```

Comment:
Should link to the correct modules/functions. Same for the args below.
```suggestion
    The lattice size is the parameter ``m`` of a :func:`dwave_networkx.pegasus_graph` or
    :func:`dwave_networkx.chimera_graph` or :func:`dwave_networkx.zephyr_graph`, also
```